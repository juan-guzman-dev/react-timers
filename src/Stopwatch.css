.Stopwatch {
    margin-left: 30px;
    margin-right: 30px;
    border: 2px solid grey;
    border-radius: 4px;
    padding: 20px;
    width: 400px;
    background-color: rgb(22, 27, 31);
    box-shadow: 0 3px 6px rgb(12, 12, 12);
  }
  
  .Stopwatch-header {
    font-size: 40px;
    font-weight: bold;
  }
  
  .Stopwatch button {
    padding: 12px 32px;
    font-size: 20px;
  }
  .Stopwatch-display {
    padding: 40px 0;
    font-size: 48px;
  }