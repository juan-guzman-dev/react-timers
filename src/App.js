import React from 'react';
import './App.css';
import Stopwatch from './Stopwatch';
import Countdown from './Countdown';
import UserForm from './UserForm';

function App() {
  return (
    <div className="App">
      <div className="App-title">Timers Demo</div>
      <div className="Timers">
        <Stopwatch />
        <Countdown />
        {/* <UserForm /> */}
      </div>

    </div>
  );
}

export default App;
