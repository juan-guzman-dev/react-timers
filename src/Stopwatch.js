import React, { useState, useEffect, useRef } from 'react';
import "./Stopwatch.css";

const Stopwatch = () => {
    const INITIAL_STATE = {
        timerOn: false, // boolean value for if the timer is on
        timerStart: 0, // the Unix Epoch (ms after 1970) time when the timer was started (or the past projected start time if the timer is resumed)
        timerTime: 0 // total time (ms) that the timer has been running since start/reset
    };

    const [timer, setTimer] = useState(INITIAL_STATE);
    const timerId = useRef()
    const { timerTime, timerOn } = timer;

    useEffect(() => {
        if (timerOn) {
            timerId.current = setInterval(() => {
                // update timerTime every 10ms
                setTimer(timer => {
                    return {
                        ...timer,
                        timerTime: Date.now() - timer.timerStart
                    }
                });
            }, 10);
        }
        return () => {
            clearInterval(timerId.current)
        }
    }, [timerOn])

    const startTimer = () => {
        setTimer(timer => {
            return {
                ...timer,
                timerOn: true,
                timerTime: timer.timerTime,
                timerStart: Date.now() - timer.timerTime
            }
        });
    };

    const stopTimer = () => {
        setTimer(timer => {
            return {
                ...timer,
                timerOn: false
            }
        });
        clearInterval(timerId.current);
    };

    const resetTimer = () => {
        setTimer(timer => {
            return {
                ...timer,
                timerStart: 0,
                timerTime: 0
            }
        });
    };

    let centiseconds = ("0" + (Math.floor(timerTime / 10) % 100)).slice(-2);
    let seconds = ("0" + (Math.floor(timerTime / 1000) % 60)).slice(-2);
    let minutes = ("0" + (Math.floor(timerTime / 60000) % 60)).slice(-2);
    let hours = ("0" + Math.floor(timerTime / 3600000)).slice(-2);

    return (
        <div className="Stopwatch">
            <div className="Stopwatch-header">Stopwatch</div>
            <div className="Stopwatch-display">
                {hours} : {minutes} : {seconds} : {centiseconds}
            </div>
            {timerOn === false && timerTime === 0 && (
                <button onClick={startTimer}>Start</button>
            )}
            {timerOn === true && (
                <button onClick={stopTimer}>Stop</button>
            )}
            {timerOn === false && timerTime > 0 && (
                <button onClick={startTimer}>Resume</button>
            )}
            {timerOn === false && timerTime > 0 && (
                <button onClick={resetTimer}>Reset</button>
            )}
        </div>
    );
}

export default Stopwatch;