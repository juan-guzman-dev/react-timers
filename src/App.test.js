import { render, screen } from '@testing-library/react';
import App from './App';
import Stopwatch from './Stopwatch';

// smoke test
it("renders without crashing", function () {
  render(<App />);
});

// snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(<Stopwatch />);
  expect(asFragment()).toMatchSnapshot();
});