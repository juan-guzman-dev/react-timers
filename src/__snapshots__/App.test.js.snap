// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Stopwatch"
  >
    <div
      class="Stopwatch-header"
    >
      Stopwatch
    </div>
    <div
      class="Stopwatch-display"
    >
      00 : 00 : 00 : 00
      <button>
        Start
      </button>
    </div>
  </div>
</DocumentFragment>
`;
